File: server/lib/SocketIO/SocketIO.go
package SocketIO

import (
	"fmt"
	"log"
	"net/http"
	"time"

	"github.com/google/uuid"
	"github.com/gorilla/websocket"
)

var upgrader = websocket.Upgrader{
	ReadBufferSize:  1024,
	WriteBufferSize: 1024,
	CheckOrigin: func(r *http.Request) bool {
		return true
	},
}

type ID_t uint32
type OnConnectHandler func(ID ID_t) error
type OnDisconnectHandler func()
type OnErrorHandler func(err error) bool
type EventsCall func(ID ID_t, msg SocketMessageEvent)

type SocketMessageEvent struct {
	EventKey string   `json:"eventKey"`
	Message  string   `json:"message"`
	To       string   `json:"to"`
	Except   []string `json:"except"`
}

type ClientsMap map[ID_t]Client

type SocketIO struct {
	clients      ClientsMap
	onConnect    OnConnectHandler
	onDisconnect OnDisconnectHandler
	onError      OnErrorHandler
	events       map[string]EventsCall
}

func New() SocketIO {
	return SocketIO{
		clients:      make(ClientsMap),
		onConnect:    func(ID ID_t) error { return nil },
		onDisconnect: func() {},
		onError:      func(err error) bool { return true },
		events:       make(map[string]EventsCall),
	}
}

func (s *SocketIO) OnConnect(callback OnConnectHandler) {
	s.onConnect = callback
}

func (s *SocketIO) OnDisconnect(callback OnDisconnectHandler) {
	s.onDisconnect = callback
}

func (s *SocketIO) OnError(callback OnErrorHandler) {
	s.onError = callback
}

func (s *SocketIO) On(eventId string, callback EventsCall) {
	s.events[eventId] = callback
}

func (s *SocketIO) Emit(eventId string, message string) {
	for _, v := range s.clients {
		v.Conn.WriteJSON(SocketMessageEvent{
			EventKey: eventId,
			Message:  message,
		})
	}
}

func (s *SocketIO) To(ID ID_t) SocketEmiter {
	res := NewSocketEmiter(s)
	res.To(ID)
	return res
}

func (s *SocketIO) Except(ID ID_t) SocketEmiter {
	res := NewSocketEmiter(s)
	res.Except(ID)
	return res
}

func (s *SocketIO) Handle(w http.ResponseWriter, r *http.Request) {
	conn, err := upgrader.Upgrade(w, r, nil)
	if err != nil {
		s.onError(err)
		return
	}
	var ID ID_t
	{
		currentTimestamp := time.Now().UnixNano() / int64(time.Microsecond)
		uniqeID := uuid.New().ID()
		ID = ID_t(uint32(currentTimestamp) + uniqeID)
	}

	s.clients[ID] = NewClient(conn)
	defer s.onDisconnect()
	defer delete(s.clients, ID)
	defer conn.Close()

	s.To(ID).Emit("connected", fmt.Sprint(ID))
	s.onConnect(ID)

	echo(conn, s, ID)
}

func echo(conn *websocket.Conn, s *SocketIO, ID ID_t) {
	for {
		socketData := &SocketMessageEvent{}
		err := conn.ReadJSON(socketData)
		if err != nil {
			break
		}

		call, exits := s.events[socketData.EventKey]

		if exits {
			call(ID, *socketData)
		} else {
			log.Printf("%s doesn't exist\n", socketData.EventKey)
		}
	}
}

File: client/src/lib/SocketIO.ts
export interface SocketMessageRequest {
	eventKey: string;
	to?: string;
	message?: string;
	data?: Object | string;
	except: string[];
}
export type SocketIOListener = (message?: string, ev?: MessageEvent) => void;
export type ID_t = string;

export class SocketEmiter {
    #to?: string;
    #except: string[];
    constructor(public readonly socketClient: SocketIO) {
        this.#to = undefined;
        this.#except = [];
    }
    public to(ID: ID_t): SocketEmiter {
        this.#to = ID;
        return this;
    }
    public except(ID: ID_t): SocketEmiter {
        this.#except.push(ID);
        return this;
    }
    public emit(eventKey: string, message?: string) {
        this.socketClient.socket.send(
            JSON.stringify({
                eventKey,
                message,
                to: this.#to,
            } as SocketMessageRequest)
        );
    }
}

export default class SocketIO {
    socket: WebSocket;
    #listeners: Record<string, SocketIOListener> = {};
    #ID: ID_t = "";

    constructor(public readonly url: URL | string) {
        this.socket = new WebSocket(url);

        this.#listeners["connected"] = (msg, ev) => {
            this.#ID = msg as any;
        };

        this.socket.onmessage = (ev) => {
            const res: SocketMessageRequest = JSON.parse(ev.data);
            if (this.#listeners[res.eventKey]) {
                this.#listeners[res.eventKey](res.message, ev);
            }
        };
    }

    get ID() {
        return this.#ID;
    }

    public close() {
        this.socket.close();
    }

    public onConnect(call: SocketIOListener) {
        this.#listeners["connected"] = (msg, ev) => {
            this.#ID = msg as any;
            call(msg, ev);
        };
    }

    public onError(call: (ev: Event) => any) {
        this.socket.onerror = call;
    }

    public onClose(call: (ev: CloseEvent) => any) {
        this.socket.onclose = (ev) => {
            call(ev);
        };
    }

    public emit(eventKey: string, message?: string) {
        this.socket.send(
            JSON.stringify({
                eventKey,
                message,
                except: [],
            } as SocketMessageRequest)
        );
    }

    public to(ID: ID_t): SocketEmiter {
        const res = new SocketEmiter(this);
        res.to(ID);
        return res;
    }

    public broadcast(): SocketEmiter {
        const res = new SocketEmiter(this);
        res.except(this.ID);
        return res;
    }

    public except(ID: ID_t): SocketEmiter {
        const res = new SocketEmiter(this);
        res.except(ID);
        return res;
    }

    public on(eventId: string, call: SocketIOListener) {
        this.#listeners[eventId] = call;
    }
}
